### PR TITLE
feat(views): view extensions are handled as normal views

### PR DESCRIPTION
**Pull request by jdalsem**
This allows for view related hooks to be triggered and to extend
extensions of views

- [x] docs
- [x] tests
- [x] check developer tools view wrapping, particularly with extended CSS/JS views
- [x] check that simplecache minify won't process each individual CSS extension in elgg.css

### CODE REVIEW

**Review by jeabakker — APPROVED** [review]

**Review by mrclay — CHANGES_REQUESTED** [review]
This is a big change in how views are processed. Let's get some more opinions on this. Also needs tests and probably views docs update.

File: engine/classes/Elgg/ViewsService.php [hunk 1]
```diff
@@ -363,6 +364,16 @@ public function renderView($view, array $vars = [], $viewtype = '', $issue_missi
 		if (strpos($view, '..') !== false) {
 			return '';
 		}
+		
+		// check for extension deadloops
+		if (!isset($vars[self::EXTENSION_TREE_KEY])) {
```
Comment (mrclay) [inline engine/classes/Elgg/ViewsService.php]: If we can avoid passing this data through $vars, we should. We could add another argument to renderView.

**Comment by mrclay** [comment]
- [ ] docs
- [ ] tests
- [ ] check developer tools view wrapping, particularly with extended CSS/JS views
- [ ] check that simplecache minify won't process each individual CSS extension in elgg.css

@Elgg/core @hypeJunction any thoughts on this? The performance penalty is 2 additional hooks per each extension rendered, but this does make view extension work like a new Elgg dev might expect.

**Review by hypeJunction — COMMENTED** [review]
Can't you just add a new hook to filter the stack before rendering starts? That will allow you to add, remove and reorder views without all o these hooks hoops

**Comment by mrclay** [comment]
> That will allow you to add, remove and reorder views

That addresses a different problem. This addresses the issue that extensions don't have their own hooks.

**Comment by hypeJunction** [comment]
Never had any such use case, so no opinion here, but seems overly complicated

**Comment by mrclay** [comment]
With this, view wrapping would now include extensions, so this seems a big win for debugging.

**Comment by mrclay** [comment]
@jdalsem @jeabakker What uses cases do you have in mind? This obviously hasn't been a big issue for Elgg and I'd lean toward keeping the status quo just because of the uncertainty of how it would affect existing view extensions in the wild.

**Comment by jdalsem** [comment]
Good to see that you can think of a very good use case yourself (the view wrapping debug feature). In the past we have had issue with unextendable extensions... Recently i wanted to use a view_vars hook to prevent output of a view (without the need to take over the view). This was the user language setting view which is an extension of user/settings view. The only way to prevent the view to output was to override a view. I wished it wasn't. So i fixed it. This PR now brings consistency to this undocumented limitation of the views system.

Will update docs if needed and change the extension recursion prevention to use an internal function param and will write test to check if extensions are extendable and if recursion is prevented.

>  I'd lean toward keeping the status quo 
There is no need to be scared

**Comment by jdalsem** [comment]
needed a small fix in developer tools to fix wrapping inside css/js

minify is not using the views system, it hooks on simplecache generation during cache creation

**Comment by jdalsem** [comment]
fixed 2 issues with unextendView and extendView in the meantime

**Comment by jdalsem** [comment]
For me this is now done and ready to go in

**Review by hypeJunction — COMMENTED** [review]
IMO needs tests for the use case you are adding this for, hooks to be precise

**Comment by jdalsem** [comment]
> IMO needs tests for the use case you are adding this for, hooks to be precise

don't see why... hooks being triggered on views is tested somewhere else. I added tested to see if the extension are used as a regular view. I can assume that the hooks will work as expected.

**Comment by jdalsem** [comment]
the use case is that extensions should use elgg_view... that is what is tested

File: engine/classes/Elgg/ViewsService.php [hunk 2]
```diff
@@ -511,9 +534,14 @@ public function viewExists($view, $viewtype = '', $recurse = true) {
 	public function extendView($view, $view_extension, $priority = 501) {
 		$view = self::canonicalizeViewName($view);
 		$view_extension = self::canonicalizeViewName($view_extension);
+		
+		if ($view === $view_extension) {
+			// do not allow direct extension on self with self
```
Comment (mrclay) [inline engine/classes/Elgg/ViewsService.php]: At least deserves a logged error.

File: engine/classes/Elgg/ViewsService.php [hunk 3]
```diff
@@ -561,8 +589,11 @@ public function unextendView($view, $view_extension) {
 		if (!isset($this->extensions[$view])) {
 			return false;
 		}
+		
+		$extensions = $this->extensions[$view];
+		unset($extensions[self::BASE_VIEW_PRIORITY]); // we do not want the base view to be removed from the list
```
Comment (mrclay) [inline engine/classes/Elgg/ViewsService.php]: What does this prevent? I guess I'm not getting it.

File: mod/developers/start.php [hunk 4]
```diff
@@ -184,6 +184,10 @@ function developers_wrap_views($hook, $type, $result, $params) {
 	if (elgg_get_viewtype() != "default") {
 		return;
 	}
+	
+	if (stristr(current_page_url(), elgg_normalize_url('cache/'))) {
```
Comment (mrclay) [inline mod/developers/start.php]: We could check this before registering the handler but this is fine.

**Comment by mrclay** [comment]
> I added tests to see if the extension are used as a regular view. I can assume that the hooks will work as expected

SGTM.

**Review by mrclay — APPROVED** [review]

File: engine/classes/Elgg/ViewsService.php (hunk 2 above)
Comment (jdalsem) [reply engine/classes/Elgg/ViewsService.php]: yep

File: engine/classes/Elgg/ViewsService.php (hunk 3 above)
Comment (jdalsem) [reply engine/classes/Elgg/ViewsService.php]: see test case https://github.com/Elgg/Elgg/pull/11231/files#diff-62ab3d7df7906064c85f6ac437fe6128R231

currently you can remove the base view from its own extension list
just do
```
elgg_extend_view('a', 'b');
elgg_unextend_view('a', 'a');
elgg_view('a'); // output will be only b
```

**Comment by hypeJunction** [comment]
Was just going to comment on this. Can we make sure that `$vars` are available in the extension views as they were before, vis registration form.

**Comment by jdalsem** [comment]
@hypeJunction could you clarify what you mean?

**Comment by hypeJunction** [comment]
Currently view extensions have access to `$vars` with any modifications made by the views being extended.

```php
elgg_extend_view('a', 'b');

echo elgg_view('a', [
   'foo' => 'bar',
]);

// a.php
echo $vars['foo']; // prints 'bar'
$vars['foo'] = 'baz';

// b.php
echo $vars['foo']; // prints 'baz'

```

I don't think we have any tests for that.

**Comment by jeabakker** [comment]
@hypeJunction your case doesn't work anymore with the introduction of the `ViewsService::renderViewFile()` (probably since Elgg 1.9)

**Comment by hypeJunction** [comment]
I don't really need it, but noticed it when working on the registration form. Sticky form values are somehow supposed to be available to form extensions.

**Comment by hypeJunction** [comment]
Perhaps it's a good time to get rid of the public sticky form API, and do everything magically in the backend, both action and forms service should be able to handle this and we can get rid of all the boilerplate.

**Comment by jeabakker** [comment]
> I don't really need it, but noticed it when working on the registration form. Sticky form values are somehow supposed to be available to form extensions.

probably due to a prepare_form_vars function

> Perhaps it's a good time to get rid of the public sticky form API, and do everything magically in the backend, both action and forms service should be able to handle this and we can get rid of all the boilerplate.

see #11141

**Pull request merged**